apache/metron
# METRON-1801 Allow Customization of Elasticsearch Document ID

Currently, the Metron GUID is always used as the Elasticsearch document ID. As documented in [METRON-1677](https://issues.apache.org/jira/browse/METRON-1677), using a randomized UUID like Java's `UUID.randomUUID()` can negatively impact Elasticsearch performance.  This change allows a user to customize the identifier that is used by Elasticsearch when indexing documents.

A property `es.document.id` was added that sets the message field that is used to define the document ID when a message is indexed by Elasticsearch. 

* To allow Elasticsearch to define its own document id, this property should be set to a blank or empty string. The client will not set the document ID and Elasticsearch will define its own.

* In most cases allowing Elasticsearch to define the document ID is the most performant option. This is the default behavior.

* Metron versions 0.6.0 and earlier defined the document ID using the Metron GUID, which is a randomized UUID using Java's `UUID.randomUUID()`. Using a randomized UUID can negatively impact Elasticsearch indexing performance. To maintain backwards compatibility with legacy versions of Metron use the following setting.

    ```
    es.document.id = guid
    ``` 

* To use a custom document ID, create an enrichment that defines a new message field; for example one called `my_document_id`. Then use this field to set the document ID as follows. This will set the document ID to the value of the message field `my_document_id`.
    ```
    es.document.id = my_document_id
    ```

* If a message does not contain the `es.document.id` field, a warning is issued and no document ID is set by the client.

## Changes

  * The `ElasticsearchWriter` was updated to allow the document ID to be configurable.

  * A 'search by GUID' in the REST layer was implicitly using the document ID, whereas it should be using the Metron GUID.

* Search results should use the Metron GUID as the ID returned to the UI.  All IDs visible to the user should always be the Metron GUID, not the document ID.

* The MPack was updated to allow the user to define the `es.document.id` on the Metron > Config > Index Settings tab.

* The default behavior was changed to allow Elasticsearch to set the document ID.  This is the most performant option in most cases.  I updated the `Upgrading.md` doc to describe how to revert to the legacy behavior.

## Testing

1. Spin-up a development environment.  You may need to stop the PCAP and/or Profiler topology to free-up slots to allow indexing to occur.

    ```
    cd metron-deployment/development/centos6
    vagrant up
    ```

1. Open the Alerts UI and ensure that alerts are visible.  Notice that the ID listed in the table has not changed.  This will always display the Metron GUID, no matter what ID used for the document.

    ![screen shot 2018-10-02 at 5 43 54 pm](https://user-images.githubusercontent.com/2475409/46378802-c5abbe00-c66a-11e8-9a65-5ce2ad84dd59.png)

1. Click on a GUID in the table to search for a single alert.

    ![screen shot 2018-10-02 at 12 42 54 pm](https://user-images.githubusercontent.com/2475409/46378629-3b635a00-c66a-11e8-9b13-6c0f54e48dfa.png)

1. Create a meta-alert and ensure that alerts tied to the meta-alert are still discoverable by GUID.

    ![screen shot 2018-10-02 at 5 45 18 pm](https://user-images.githubusercontent.com/2475409/46378860-f3910280-c66a-11e8-8612-e18ab2e4ae03.png)

1. Open Ambari and go to Metron > Configs > Index Settings.  Notice that the default setting of `es.document.id` is blank, thus allowing Elasticsearch to define its own document ID.

    ![screen shot 2018-10-03 at 9 13 18 am](https://user-images.githubusercontent.com/2475409/46412476-a81f3880-c6ec-11e8-9b1e-38fe36c18b03.png)

1. Open Kibana and verify that indeed Elasticsearch is generated its own document IDs.  You will notice an `_id` field which has been generated by Elasticsearch.  This will be different than the UUID generated by Metron and stored as part of the document as `guid`.

    ![screen shot 2018-10-02 at 4 47 52 pm](https://user-images.githubusercontent.com/2475409/46378868-fc81d400-c66a-11e8-9981-3e62033e44ec.png)

1. Stop the indexing topologies using Ambari.

1. Login to the VM.

    ```
    vagrant ssh
    sudo su -
    ```

1. Delete the existing indices in Elasticsearch.

    ```
    curl -XDELETE http://node1:9200/bro*
    curl -XDELETE http://node1:9200/snort*
    ```

1. Open Ambari, and go to Metron > Configs > Index Settings.  Edit the "Elasticsearch Document ID Source Field" and set it to 'guid'.  This will restore the legacy behavior where the document ID is to the Metron GUID.

1. Restart the Indexing Topology in Ambari. 

1. Open the Alerts UI and ensure that alerts are visible.  Notice that the ID listed in the table has not changed.  This will always display the Metron GUID, no matter what ID used for the document.

1. Click on a GUID in the table to search for a single alert.

    ![screen shot 2018-10-02 at 12 42 54 pm](https://user-images.githubusercontent.com/2475409/46378629-3b635a00-c66a-11e8-9b13-6c0f54e48dfa.png)

1. Create a meta-alert and ensure that alerts tied to the meta-alert are still discoverable by GUID.

    ![screen shot 2018-10-02 at 5 45 18 pm](https://user-images.githubusercontent.com/2475409/46378860-f3910280-c66a-11e8-8612-e18ab2e4ae03.png)

1. Open Kibana and verify that the document ID matches the Metron GUID.

    ![screen shot 2018-10-03 at 5 58 49 pm](https://user-images.githubusercontent.com/2475409/46442312-391b0180-c737-11e8-8c5f-574eefa10090.png)

1. I would also advise running the UI e2e test suite with this change.

## Pull Request Checklist
